### PR TITLE
Adding Compass Tool

### PR DESCRIPTION
**Pull request by sanal-cohere**
Thank you for contributing to the Cohere Toolkit!

- [ ] **PR title**: "area: description"
  - Where "area" is whichever of interface, frontend, model, tools, backend, etc. is being modified. Use "docs: ..." for purely docs changes, "infra: ..." for CI changes.
  - Example: "deployment: add Azure model option"

- [ ] **PR message**: ***Delete this entire checklist*** and replace with
    - **Description:** a description of the change
    - **Issue:** the issue # it fixes, if applicable
    - **Dependencies:** any dependencies required for this change

- [ ] **Add tests and docs**: Please include testing and documentation for your changes
- [ ] **Lint and test**: Run `make lint` and `make run-tests`

**AI Description**

This PR introduces a new Python class, `CompassTool`, which extends the `BaseTool` class. This class provides functionality to interact with a Compass instance and perform operations such as creating, searching, updating, and deleting documents.

## File Walkthrough:
The changes are made in the `src/backend/tools/compass.py` file:
- New imports are added, including modules from the `compass_sdk` package.
- The `CompassTool` class is defined, inheriting from `BaseTool`.
- The `is_available` class method checks for the existence of required environment variables related to Compass configuration.
- The `__init__` method initializes the `CompassTool` class by setting the Compass URL, username, and password, either through direct arguments or environment variables.
- The `call` method serves as the entry point for using the `CompassTool`. It performs various checks to ensure that the required parameters are provided and then delegates the action to specific methods (`_create`, `_search`, `_update`, `_delete`) based on the specified action.
- The `_parse` method is responsible for parsing the input file and validating its existence.
- The `_create` method inserts a document into Compass by first parsing the input and then using the `insert_docs` method of the `CompassClient`.
- The `_search` method executes a search query on Compass and returns the top-k results.
- The `_update` method updates a file in Compass by first deleting and then recreating it.
- The `_delete` method removes a file from Compass using the `delete_document` method of the `CompassClient`.

## Summary:
The `CompassTool` class provides a comprehensive set of methods to interact with a Compass instance, offering a user-friendly way to perform common operations on Compass documents.

### CODE REVIEW

File: src/backend/tools/compass.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        # Check if action is specified
+        if not parameters.get("action", None):
+            logger.error(
+                "Compass Tool: No action specified. "
+                "No action will be taken. "
+                f"Parameters specified: {parameters}"
+            )
+            return
+        # Check if index is specified
+        if not parameters.get("index", None):
```
Comment (benrules3) [inline src/backend/tools/compass.py:56]: There is one action `list_index` that would not have index  parameter. Was skipping `list_index` deliberate?

File: src/backend/tools/compass.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        ):
+        """Initialize the Compass tool. Pass the Compass URL, username, and password 
+        as arguments or as environment variables."""
+        self.url = compass_url or os.getenv("COHERE_COMPASS_URL")
+        self.username = compass_username or os.getenv("COHERE_COMPASS_USERNAME")
+        self.password = compass_password or os.getenv("COHERE_COMPASS_PASSWORD")
+        self.parser_client = CompassParserClient(
+            parser_url=self.url + "/parse",
+            parser_config=self.username,
+            password=self.password,
```
Comment (benrules3) [inline src/backend/tools/compass.py:37]: This doesn't specify a `MetadataConfig` anywhere, which mean's were just using the default with no metadata extraction. This is perfectly fine, but just calling out we're accepting the defaults here.

**Review by benrules3 — COMMENTED** [review]
Approach is looking good 👍

File: src/backend/tools/compass.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    def __init__(
+            self,
+            compass_url=None,
+            compass_username=None,
+            compass_password=None,
+        ):
+        """Initialize the Compass tool. Pass the Compass URL, username, and password 
+        as arguments or as environment variables."""
+        self.url = compass_url or os.getenv("COHERE_COMPASS_URL")
```
[the remark this answers is not in the document]
Comment (sanal-cohere) [reply src/backend/tools/compass.py:33]: Agreed, we should also check of invalid credentials, urls etc somewhere. I added a try-except block with client.list_indexes(). Can you take a look to see if this is adequate to catch the issue?

File: src/backend/tools/compass.py (hunk 1 above)
Comment (sanal-cohere) [reply src/backend/tools/compass.py]: I added `list_indexes`, `create_index`, and the `delete_index` actions to the tool.

File: src/backend/tools/compass.py (hunk 2 above)
Comment (sanal-cohere) [reply src/backend/tools/compass.py]: I passed the parser_config and metadata_config as parameters with default values. This should allow for customization when the tool is created. Also included the ability to pass parser_config and metadata_config in the to the _parse method.

**Review by benrules3 — COMMENTED** [review]
Compass integration looks right, good updates 👍

**Review by lusmoura — APPROVED** [review]
LGTM!

**Pull request merged**